---
title: "Gene Expression Analysis"
author: "Giacomo"
date: "7/19/2021"
output: html_document
---

```{r}
library(tidyverse)
library(edgeR)
library(devtools)
library(Rsubread)
library(ggplot2)
library(rtracklayer)
library(dplyr)
library(knitr)
counts.data <- read_tsv("allcounts.tsv")
colnames(counts.data) <- str_replace(colnames(counts.data), fixed(".ReadsPerGene.out.tab"),"")
colnames(counts.data) <- str_replace(colnames(counts.data), fixed("Sample"), ("Gene_ID"))
counts.data <- counts.data %>%
  filter(Gene_ID!="N_unmapped",Gene_ID!="N_multimapping", Gene_ID!="N_noFeature", Gene_ID!="N_ambiguous" )
counts.data
  
```
The sample refers to the f01_etc and the evm is the gene id of each individual. Basically, evm.model.scaffold_1603.163 (specific gene) in sample f01_S55_L001 has 0 reads. 
```{r}
M_K <- counts.data[,c(1, 61:70, 86:95)]
M_K
M_K <- M_K[rowSums(counts.data[,-1] > 10) >= 3,]
colnames(M_K) <- str_replace(colnames(M_K), fixed("k"), ("stin"))
colnames(M_K) <- str_replace(colnames(M_K), fixed("m"), ("stto-bh"))
Light <- M_K[, c(1:6, 12:16)]
colnames(Light) <- str_replace(colnames(Light), fixed("S"), ("L"))
Dark <- M_K[, c(1, 7, 8, 10:11, 17:21)]
Dark
colnames(Dark) <- str_replace(colnames(Dark), fixed("S"), ("D"))
M_K2 <- merge(Light, Dark)
M_K2
M_K3 <- M_K2[,c(1:6, 12:16, 7, 8, 10:11, 17:20)]
M_K3
```

```{r}
vis.counts <- M_K[,-1]
vis.counts <-vis.counts[1:1000,]
cor(vis.counts)
```

```{r}
sample.description <- tibble(sample=colnames(M_K3)[-1])
sample.description
sample_seperation <- sample.description %>%
  mutate(SP = str_extract(sample, "stin|stto-bh"),
         trt = str_extract(sample, "L|D"),
         group = str_c(SP, trt, sep = "_"))

sample_seperation <- sample_seperation %>%
  mutate(SP=factor(SP),
         trt=factor(trt, levels = c("D","L")))
sample_seperation
```

```{r}
counts.matrix <- M_K3 %>%
  select(-Gene_ID)%>%
  as.matrix()
rownames(counts.matrix) <- M_K3$Gene_ID
dge.data <- DGEList(counts=counts.matrix,
                    group=sample_seperation$group)

dim(dge.data)
dge.data <- calcNormFactors(dge.data, method = ("TMM"))
dge.data$samples
plotMDS(dge.data, method = "bcv")
```

```{r}
counts.data.normal <- cpm(dge.data)
counts.data.normal.log <- cpm(dge.data, log = TRUE)
counts.data.log <- log2(M_K3[,-1] + 1)
boxplot(counts.data.log)
boxplot(counts.data.normal.log)
```
```{r}

design <- model.matrix(~SP+trt, data = sample_seperation)
rownames(design) <- sample_seperation$sample
design
dge.data <- estimateGLMTrendedDisp(dge.data, design)
dge.data <- estimateGLMTagwiseDisp(dge.data, design)
dge.data <- estimateGLMCommonDisp(dge.data, design)
plotBCV(dge.data)
```
```{r}
fit <- glmFit(dge.data, design)
design
trtL.lrt <- glmLRT(fit, coef = "trtL")
topTags(trtL.lrt)
summary(decideTestsDGE(trtL.lrt, p.value =0.01))

DEgene.trtL <- topTags(trtL.lrt,n = Inf,p.value = 0.01)$table
write.csv(DEgene.trtL,"DEgenes.trtL.csv")
DEgene.trtL.all <- topTags(trtL.lrt,n = Inf, p.value = 1)$table
write.csv(DEgene.trtL.all,"DEgenes.trtL.all.csv")

plotDE <- function(genes, dge, sample_seperation) {
  require(ggplot2)
  tmp.data <- t(log2(cpm(dge[genes,])+1))
  tmp.data <- tmp.data %>%
    as.data.frame() %>%
    rownames_to_column("sample") %>%
    left_join(sample_seperation,by="sample")
  tmp.data <- tmp.data %>%
    pivot_longer(cols=starts_with("evm"), values_to = "log2_cpm", names_to = "gene")
  pl <- ggplot(tmp.data,aes(x= SP,y=log2_cpm,fill= trt))
  pl <- pl + facet_wrap( ~ gene)
  pl <- pl + ylab("log2(cpm)") + xlab("treatment")
  pl <- pl + geom_boxplot()
  pl + theme(axis.text.x  = element_text(angle=45, vjust=1,hjust=1))
}
plotDE (rownames(DEgene.trtL)[1:9],dge.data, sample_seperation)
plotDE ("evm.model.scaffold_34.25987", dge.data, sample_seperation)
```

Ex. Bottom right. That particular gene is expressed more highly in light in both of the species. Also would you want to look at the group so trt combined with individual species to see that a particular gene in a particular species is expressed more highly in that species than the other? That, that gene is expressed more highly in light in both species. 
evm.modelscaffold_34.25987 is expressed 5*x higher in light
a negative number would suggest that the gene expressed `blank` times more in the dark. 
These genes are being expressed differently under light in germinating conditions. 


```{r}
colSums(M_K3[,-1])
```
```{r}
design1 <- model.matrix(~SP*trt, data = sample_seperation)
rownames(design1) <- sample_seperation$sample
design1
dge.data <- estimateGLMTrendedDisp(dge.data, design1)
dge.data <- estimateGLMTagwiseDisp(dge.data, design1)
dge.data <- estimateGLMCommonDisp(dge.data, design1)
plotBCV(dge.data)

fit <- glmFit(dge.data, design1)
sptrt.lrt <- glmLRT(fit, coef = "SPstto-bh:trtL")
topTags(sptrt.lrt)
summary(decideTestsDGE(sptrt.lrt, p.value =0.01))

DEgene.sptrt <- topTags(sptrt.lrt,n = Inf,p.value = 0.01)$table
write.csv(DEgene.sptrt,"DEgenes.sptrt.csv")
DEgene.sptrt.all <- topTags(sptrt.lrt,n = Inf, p.value = 1)$table
write.csv(DEgene.sptrt.all,"DEgenes.sptrt.all.csv")

plotDE <- function(genes, dge, sample_seperation) {
  require(ggplot2)
  tmp.data <- t(log2(cpm(dge[genes,])+1))
  tmp.data <- tmp.data %>%
    as.data.frame() %>%
    rownames_to_column("sample") %>%
    left_join(sample_seperation,by="sample")
  tmp.data <- tmp.data %>%
    pivot_longer(cols=starts_with("evm"), values_to = "log2_cpm", names_to = "gene")
  pl <- ggplot(tmp.data,aes(x= SP,y=log2_cpm,fill= trt))
  pl <- pl + facet_wrap( ~ gene)
  pl <- pl + ylab("log2(cpm)") + xlab("species")
  pl <- pl + geom_boxplot()
  pl + theme(axis.text.x  = element_text(angle=45, vjust=1,hjust=1))
}
plotDE (rownames(DEgene.sptrt)[1:9],dge.data, sample_seperation)
plotDE ("evm.model.scaffold_148.17662", dge.data, sample_seperation)
```
This is showing that genotype and species are having an effect on gene expression levels. 


Each species for the poarticular gene has a different interaction with light. 
What this is then showing us is that both genotype and treatment are having an impact on the expression levels. 
Treatment and genotype dictate expression levels.

```{r}
counts.data <- read_tsv("allcounts.tsv")
counts.data
ggplot(data = counts.data, mapping = aes(y = colsum)) +
  geom_col()
```

This is the start of the interaction GO analysis

```{r}
library(tidyverse)
library(goseq)
library(rtracklayer)
library(GenomicRanges)
library(Biostrings)
```

```{r}
SPtrt <- DEgene.sptrt %>%
  rownames_to_column()
colnames(SPtrt) <- str_replace(colnames(SPtrt), fixed("rowname"), "GeneID")

file.info <- read_tsv("Sdiv2Athal.blastp_20210714.out", col_names = FALSE)
colnames(file.info)[1] <- "GeneID"
colnames(file.info)[2] <- "GOTerms"
file.info
uniq.blast.results <- file.info %>%
    group_by(X12, GeneID) %>%
    filter(rank(dplyr::desc(X12), ties.method = "first") == 1) 
uniq.blast.results


GODEgene.trt <- inner_join(SPtrt, file.info)
GODEgene.trt 
```

```{r}
go.terms <- read_tsv("Sdiv_GoTerms.tsv")
go.terms

expressed.genes <- M_K3
colnames(expressed.genes) <- str_replace(colnames(expressed.genes), fixed("Gene_ID"), "GeneID")
expressed.genes


txdb <- makeTxDbFromGFF("streptanthus_jamg.gene_structures_post_PASA_updates.97994.gff3")
# then collect the exons per gene id
exons.list.per.gene <- exonsBy(txdb,by="gene")
# then for each gene, reduce all the exons to a set of non overlapping exons, calculate their lengths (widths) and sum then
exonic.gene.sizes <- sum(width(reduce(exons.list.per.gene)))
genething <- data.frame(as.data.frame(exonic.gene.sizes))
gene.length <- genething %>%
  rownames_to_column()
colnames(gene.length) <- str_replace(colnames(gene.length), fixed("rowname"), ("GeneID"))
colnames(gene.length) <- str_replace(colnames(gene.length), fixed("exonic.gene.sizes"), ("Length"))
gene.length <- gene.length %>%
  mutate(GeneID=str_replace(GeneID, "TU", "model"))
gene.length



gene.lengths.vector <- gene.length$Length[gene.length$GeneID %in% expressed.genes$GeneID]
gene.lengths.vector
names(gene.lengths.vector) <- gene.length$GeneID[gene.length$GeneID %in% expressed.genes$GeneID]
head(gene.lengths.vector)
expressed.genes.match <- expressed.genes[expressed.genes$GeneID %in% names(gene.lengths.vector),]
expressed.genes.match
```
```{r}
go.list <- strsplit(go.terms$GOs, split=",")
names(go.list) <- go.terms$GeneID
head(go.list)
```
```{r}
DE.trt <- expressed.genes.match$GeneID %in% SPtrt.all$GeneID
names(DE.trt) <- expressed.genes.match$GeneID
head(DE.trt)
DE.trt <- as.numeric(DE.trt) 
head(DE.trt)
sum(DE.trt) 
```

```{r}
nullp.result <- nullp(DEgenes = DE.trt,bias.data = gene.lengths.vector)

rownames(nullp.result) <- names(gene.lengths.vector) #because of a bug in nullp()
GO.out <- goseq(pwf = nullp.result, gene2cat = go.list,test.cats=("GO:BP"))


GO.out[GO.out$over_represented_pvalue < 0.05,]
GO.out
```
```{r}
write.table(GO.out[GO.out$over_represented_pvalue < 0.05,1:2],row.names=FALSE,file="GO_terms3.txt", quote = FALSE,col.names = FALSE)

```

This is the end of the interaction GO analysis
